search/populate.py

```diff
@@ -0,0 +1,81 @@
+import logging
+from io import BytesIO
+from typing import Union
+
+from flask import Blueprint, Response
+
+import requests
+from elasticsearch import Elasticsearch
+from lxml import etree
+
+from search.constants import (CONTENT_SERVICES_TO_INDEX, ELASTICSEARCH_HOSTS,
+                              GATEWAY_URL)
+from search.utils import strip_slashes
+
+LOGGER = logging.getLogger(__name__)
+
+
+def request_data(url: str) -> Union[requests.Response, None]:
+    try:
+        response = requests.get(url)
+        response.raise_for_status()
+        return response
+    except requests.HTTPError as error:
+        LOGGER.exception('%s %s: Unable to retrieve content from %s.',
+                         error.response.reason,
+                         error.response.status_code,
+                         url)
+    return None
+
+
+def index_data(index: str, doc_type: str, id: str, body: dict) -> None:
+    try:
+        search = Elasticsearch(hosts=ELASTICSEARCH_HOSTS)
+        search.index(index=index, doc_type=doc_type, id=id, body=body)
+        LOGGER.debug('Document %s saved at /%s/%s/%s', id, index, doc_type, id)
+    except ConnectionError as error:
+        LOGGER.exception('Unable to index /%s/%s/%s: %s', index, doc_type, id, error)
+
+
+def get_populate_blueprint() -> Blueprint:
+    blueprint = Blueprint('populate', __name__)
+
+    @blueprint.route('/populate', methods=['POST'])
+    def populate() -> Response:
+        """
+        Route for triggering indexing of known paths. Will be used as a proof of
+        concept. This should be removed before MVP release.
+        """
+        gateway = strip_slashes(GATEWAY_URL)
+        document_count = 0
+
+        for path in CONTENT_SERVICES_TO_INDEX:
+            url = f'{gateway}/{path}/items'
+            response = request_data(url)
+            if not response:
+                continue
+
+            root = etree.parse(BytesIO(response.content)).getroot()
+            for item in root.iter('{http://libero.pub}item-ref'):
+                item_url = url + f'/{item.attrib.get("id")}/versions/latest'
+                response = request_data(item_url)
+                if not response:
+                    continue
+
+                item_root = etree.parse(BytesIO(response.content)).getroot()
+                document = {}
+                for tag in ['id', 'service']:
+                    element = item_root.find(f'libero:meta/libero:{tag}',
+                                             namespaces={'libero': 'http://libero.pub'})
+                    if element.text:
+                        document[tag] = element.text
+
+                if document and 'id' in document:
+                    index_data(
+                        index='articles', doc_type='article', id=document['id'], body=document
```

Comment:
Probably shouldn't hard-code `index`/`doc_type`. `index` should be `content`/`items` if anything; `doc_type` maybe the service name? (Not sure what the implications are.)

Reply:
This is where we need to have discussion about elastic search persistence and come up with a naming convention.